frontend/src/pages/CompareV2.tsx

```diff
@@ -158,6 +196,94 @@ function getRunArtifacts(runs: ApiRunDetail[], mlmdPackages: MlmdPackage[]): Run
   });
 }
 
+interface VisualizationPlaceholderProps {
+  metricsTabText: string;
+}
+
+function VisualizationPlaceholder(props: VisualizationPlaceholderProps) {
+  const { metricsTabText } = props;
+  return (
+    <div className={classes(css.visualizationPlaceholder)}>
+      <p className={classes(css.visualizationPlaceholderText)}>
+        The selected {metricsTabText} will be displayed here.
+      </p>
+    </div>
+  );
+}
+
+interface MetricsDropdownProps {
+  filteredRunArtifacts: RunArtifacts[];
+  metricsTabText: string;
+}
+
+function MetricsDropdown(props: MetricsDropdownProps) {
+  const { filteredRunArtifacts, metricsTabText } = props;
+  const [selectedItem, setSelectedItem] = useState<SelectedItem>({ itemName: '', subItemName: '' });
+
+  const dropdownItems: DropdownItem[] = [];
+  for (const x of filteredRunArtifacts) {
+    const runName = x.run.run?.name;
+    if (runName) {
+      const subItems: DropdownSubItem[] = [];
+      for (const y of x.executionArtifacts) {
+        const executionName = y.execution
+          .getCustomPropertiesMap()
+          .get('display_name')
+          ?.getStringValue();
+        if (executionName) {
+          const executionLinkedArtifacts: DropdownSubItem[] = [];
+          for (const z of y.linkedArtifacts) {
+            const artifactName = z.event
+              .getPath()
+              ?.getStepsList()[0]
+              .getKey();
+            if (artifactName) {
+              executionLinkedArtifacts.push({
+                name: executionName,
+                secondaryName: artifactName,
+              } as DropdownSubItem);
+            } else {
+              logger.warn(
+                `Failed to fetch the display name of the artifact with the following ID: ${z.artifact.getId()}`,
+              );
+            }
+          }
+          subItems.push(...executionLinkedArtifacts);
+        } else {
+          logger.warn(
+            `Failed to fetch the display name of the execution with the following ID: ${y.execution.getId()}`,
+          );
+        }
+      }
```

Comment:
1. It will be helpful to extract this logic to a function to facilitate readability because this is highly nested code block.
2. Or, you can consider the following approach to simplify implementation:

```
filteredRunArtifacts.map((runArtifact) => {
     return runArtifact.executionArtifacts.map((executionArtifact) => {
          return executionArtifact.linkedArtifacts.map((linkedArtifact) => {
                  // Construct DropdownSubItem
                  return dropdownSubItem[]
          }).flat()
     }).flat()
}).flat()
```

Reply:
I chose the first choice because my current approach filters out artifacts for which I cannot retrieve the display name. I tried to reduce some redundancy and add some clarifying comments to the large code block as well, see here: https://github.com/kubeflow/pipelines/pull/7943/commits/205e01b762162a3de30f0367457372bac987c252.

I can use the second method if I leave all of the artifacts in, related to this discussion: https://github.com/kubeflow/pipelines/pull/7943#discussion_r906699986. Do you have any opinion there?

Reply:
It sounds good on going with the first option, thank you Chris!